### ISSUE TITLE
5.1 regression: constraints of generic tuple types are sometimes order-sensitive

### ISSUE THREAD

**Issue by alecgibson** [microsoft/TypeScript#54495]
# Bug Report

### 🔎 Search Terms

- 5.1
- tuple
- array
- spread

Possibly related:
- https://github.com/microsoft/TypeScript/issues/54048
- https://github.com/microsoft/TypeScript/issues/54302

### 🕗 Version & Regression Information

- This changed between versions v5.0.4 and v5.1.3

### ⏯ Playground Link

I was unable to reproduce: if you could please steer me in the right direction I'm happy to dig more, but I'm a bit stuck on how to debug this

### 💻 Code

We have an issue using [`io-ts`](https://github.com/gcanti/io-ts). If this is a bug with that library, I'll close this here, but I don't *think* it is.

The TypeScript compiler is complaining about a number of excerpts from the `d.ts` file, where the compiler doesn't seem to be making sense.

Here's an example error:

```
node_modules/io-ts/lib/index.d.ts:1151:24 - error TS2344: Type 'CS' does not satisfy the constraint '[Mixed, Mixed, Mixed[]]'.
  Type '[Mixed, Mixed, ...Mixed[]]' is not assignable to type '[Mixed, Mixed, Mixed[]]'.
    Target allows only 3 element(s) but source may have more.

1151 ) => TaggedUnionC<Tag, CS>
```

The code it's referring to is here:

```ts
export interface TaggedUnionC<Tag extends string, CS extends [Mixed, Mixed, ...Array<Mixed>]>
  extends TaggedUnionType<Tag, CS, TypeOf<CS[number]>, OutputOf<CS[number]>, unknown> {}

export declare const taggedUnion: <Tag extends string, CS extends [Mixed, Mixed, ...Mixed[]]>(
  tag: Tag,
  codecs: CS,
  name?: string
) => TaggedUnionC<Tag, CS>
```

As you can see, the compiler seems to be collapsing the `[Mixed, Mixed, ...Array<Mixed>]` to `[Mixed, Mixed, Mixed[]]`, which is incorrect. It then doesn't like `CS` extending `[Mixed, Mixed, ...Mixed[]]`, which it considers incompatible with its incorrect collapsing.

Weirdly, if I tweak the `d.ts` file by hand, the error goes away:

```diff
- export interface TaggedUnionC<Tag extends string, CS extends [Mixed, Mixed, ...Array<Mixed>]>
+ export interface TaggedUnionC<Tag extends string, CS extends [Mixed, Mixed, ...Mixed[]]>
```

### 🙁 Actual behavior

Compilation errors

### 🙂 Expected behavior

Successful compilation

**Comment by RyanCavanaugh** [comment]
I'm not able to reproduce this error. Can you post a concrete example?

**Comment by alecgibson** [comment]
@RyanCavanaugh I tried to put together an MWE but I couldn't figure it out.

For some reason the error only happens with io-ts as a dependency of a dependency. The middle dependency compiles just fine with TS 5.1 and the issue only happens downstream with the transpiled JS and declaration files being used in 5.1.

Do you have any advice on how to get more information from TS, or a good guess at what might be the actual issue for me to narrow my search?

**Comment by Andarist** [comment]
I'd recommend bisecting your situation by removing code from your dependencies, once you minimize this to a minimal repro you might want to just commit `node_modules` and share that as a repro case.

**Comment by alecgibson** [comment]
Okay @RyanCavanaugh @Andarist I've got a [minimal repro repo](https://github.com/alecgibson/ts-spread-bug).

Turns out, nothing to do with our code, so either a TS bug, or a bug with another repo (still looks like TS to my uneducated eye, though).

The repro case — in the end — was laughably simple:

```ts
import 'mongodb';
import 'io-ts';
```

☝🏼 this fails to build with a bare-bones `tsconfig.json` on TypeScript v5.1.3 with the compilation error I mentioned above.

It builds fine on TypeScript v5.0.4.

**Comment by alecgibson** [comment]
Please let me know if this is actually a bug with `mongodb` or `io-ts`, and I'll move there, but building a project that just depends on one of these  libraries works just fine:

```diff
import 'mongodb';
- import 'io-ts';
```

or 

```diff
- import 'mongodb';
import 'io-ts';
```

both work

**Comment by alecgibson** [comment]
Also here's the full error dump:

```
❯ npm run build

> ts-spread-bug@1.0.0 build
> tsc --pretty

node_modules/io-ts/lib/index.d.ts:771:21 - error TS2344: Type 'CS' does not satisfy the constraint 'Any[]'.
  Type '[Mixed, Mixed, Mixed[]]' is not assignable to type 'Any[]'.
    Type 'Mixed | Mixed[]' is not assignable to type 'Any'.
      Type 'Mixed[]' is not assignable to type 'Type<any, any, any>'.

771   extends UnionType<CS, TypeOf<CS[number]>, OutputOf<CS[number]>, unknown> {}
                        ~~

node_modules/io-ts/lib/index.d.ts:771:32 - error TS2344: Type 'CS[number]' does not satisfy the constraint 'Any'.
  Type 'Mixed | Mixed[]' is not assignable to type 'Type<any, any, any>'.
    Type 'Mixed[]' is missing the following properties from type 'Type<any, any, any>': name, is, validate, encode, and 7 more.

771   extends UnionType<CS, TypeOf<CS[number]>, OutputOf<CS[number]>, unknown> {}
                                   ~~~~~~~~~~

node_modules/io-ts/lib/index.d.ts:771:54 - error TS2344: Type 'CS[number]' does not satisfy the constraint 'Any'.

771   extends UnionType<CS, TypeOf<CS[number]>, OutputOf<CS[number]>, unknown> {}
                                                         ~~~~~~~~~~

node_modules/io-ts/lib/index.d.ts:799:5 - error TS2344: Type 'CS' does not satisfy the constraint 'Any[]'.
  Type '[Mixed, Mixed, Mixed[]]' is not assignable to type 'Any[]'.
    Type 'Mixed | Mixed[]' is not assignable to type 'Any'.
      Type 'Mixed[]' is not assignable to type 'Type<any, any, any>'.

799     CS,
        ~~

node_modules/io-ts/lib/index.d.ts:807:48 - error TS2344: Type 'CS[2]' does not satisfy the constraint 'Any'.
  Type 'Mixed[]' is missing the following properties from type 'Type<any, any, any>': name, is, validate, encode, and 7 more.

807       ? TypeOf<CS[0]> & TypeOf<CS[1]> & TypeOf<CS[2]>
                                                   ~~~~~

node_modules/io-ts/lib/index.d.ts:824:54 - error TS2344: Type 'CS[2]' does not satisfy the constraint 'Any'.
  Type 'Mixed[]' is missing the following properties from type 'Type<any, any, any>': name, is, validate, encode, and 7 more.

824       ? OutputOf<CS[0]> & OutputOf<CS[1]> & OutputOf<CS[2]>
                                                         ~~~~~

node_modules/io-ts/lib/index.d.ts:846:58 - error TS2344: Type '[A, B, C, D, E]' does not satisfy the constraint '[Mixed, Mixed, Mixed[]]'.
  Source has 5 element(s) but target allows only 3.

846 >(codecs: [A, B, C, D, E], name?: string): IntersectionC<[A, B, C, D, E]>
                                                             ~~~~~~~~~~~~~~~

node_modules/io-ts/lib/index.d.ts:850:18 - error TS2344: Type '[A, B, C, D]' does not satisfy the constraint '[Mixed, Mixed, Mixed[]]'.
  Source has 4 element(s) but target allows only 3.

850 ): IntersectionC<[A, B, C, D]>
                     ~~~~~~~~~~~~

node_modules/io-ts/lib/index.d.ts:854:18 - error TS2344: Type '[A, B, C]' does not satisfy the constraint '[Mixed, Mixed, Mixed[]]'.
  Type at position 2 in source is not compatible with type at position 2 in target.
    Type 'C' is not assignable to type 'Mixed[]'.
      Type 'Type<any, any, unknown>' is missing the following properties from type 'Mixed[]': length, pop, push, concat, and 29 more.

854 ): IntersectionC<[A, B, C]>
                     ~~~~~~~~~

node_modules/io-ts/lib/index.d.ts:1139:32 - error TS2344: Type 'CS' does not satisfy the constraint 'Mixed[]'.
  Type '[Mixed, Mixed, Mixed[]]' is not assignable to type 'Mixed[]'.
    Type 'Mixed | Mixed[]' is not assignable to type 'Mixed'.
      Type 'Mixed[]' is missing the following properties from type 'Type<any, any, unknown>': name, is, validate, encode, and 7 more.

1139   extends TaggedUnionType<Tag, CS, TypeOf<CS[number]>, OutputOf<CS[number]>, unknown> {}
                                    ~~

node_modules/io-ts/lib/index.d.ts:1139:43 - error TS2344: Type 'CS[number]' does not satisfy the constraint 'Any'.
  Type 'Mixed | Mixed[]' is not assignable to type 'Type<any, any, any>'.
    Type 'Mixed[]' is not assignable to type 'Type<any, any, any>'.

1139   extends TaggedUnionType<Tag, CS, TypeOf<CS[number]>, OutputOf<CS[number]>, unknown> {}
                                               ~~~~~~~~~~

node_modules/io-ts/lib/index.d.ts:1139:65 - error TS2344: Type 'CS[number]' does not satisfy the constraint 'Any'.

1139   extends TaggedUnionType<Tag, CS, TypeOf<CS[number]>, OutputOf<CS[number]>, unknown> {}
                                                                     ~~~~~~~~~~

node_modules/io-ts/lib/index.d.ts:1151:24 - error TS2344: Type 'CS' does not satisfy the constraint '[Mixed, Mixed, Mixed[]]'.
  Type '[Mixed, Mixed, ...Mixed[]]' is not assignable to type '[Mixed, Mixed, Mixed[]]'.
    Target allows only 3 element(s) but source may have more.

1151 ) => TaggedUnionC<Tag, CS>
                            ~~


Found 13 errors in the same file, starting at: node_modules/io-ts/lib/index.d.ts:771
```

**Comment by Andarist** [comment]
This can be bisected to this PR https://github.com/microsoft/TypeScript/pull/53672 , cc @ahejlsberg 

That being said, this isn't quite a minimal repro case 😉 `mongodb` and `io-ts` have a lot of types and it might take a while for somebody to actually reduce this. 

Note that reordering those 2 imports also fixes the issue so it's likely a caching bug in the compiler. Things are cached for performance reasons and unfortunately, sometimes issues like this pop up and at times it's hard to fix them. Ideally, the order of declarations shouldn't matter but in practice, it sometimes might.

**Comment by Andarist** [comment]
Maybe something like: "5.1 regression: constraints of generic tuple types are sometimes order-sensitive"

**Comment by alecgibson** [comment]
Okay I spotted something as I was skimming the MongoDB type declarations, and it looks like if you comment out this [suspicious-looking type](https://github.com/mongodb/node-mongodb-native/blob/008fd6f2c74230be1b60907918eed47617c681d1/src/mongo_types.ts#L459-L466) (and things that depend on it), everything compiles just fine.

I've pushed the result to the above repo as a new branch: [`commented-deps`](https://github.com/alecgibson/ts-spread-bug/tree/commented-deps)

Is this a TypeScript bug or a MongoDB bug?

**Comment by alecgibson** [comment]
@Andarist @ahejlsberg this is still broken in `5.1.5`. Should it be fixed?

**Comment by Andarist** [comment]
It seems that the fix was supposed to be cherry-picked into 5.1 but the automated job failed, see the comment [here](https://github.com/microsoft/TypeScript/pull/54536#issuecomment-1589656627). cc @DanielRosenwasser @jakebailey

**Comment by jakebailey** [comment]
I just sent #54814 to cherry-pick it by hand. Unfortunate; I totally read that notification in my email but I figured it was already handled...
